aesd-char-driver/main.c

```diff
@@ -0,0 +1,287 @@
+
+/**
+ * @file aesdchar.c
+ * @brief Functions and data related to the AESD char driver implementation
+ *
+ * Based on the implementation of the "scull" device driver, found in
+ * Linux Device Drivers example code.
+ *
+ * @author Dan Walkes
+ * @date 2019-10-22
+ * @copyright Copyright (c) 2019
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/printk.h>
+#include <linux/types.h>
+#include <linux/cdev.h>
+#include <linux/fs.h> 
+#include "aesdchar.h"
+#include "linux/slab.h"
+#include "linux/string.h"
+int aesd_major =   0; // use dynamic major
+int aesd_minor =   0;
+
+MODULE_AUTHOR("Sudarshan Jagannathan"); /** TODO: fill in your name **/
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    struct aesd_dev *dev; /* Pointer to the aesd_dev struct*/
+    
+    PDEBUG("open");
+    
+	dev = container_of(inode->i_cdev, struct aesd_dev, cdev);   //The address of the container structure where cdev is located is returned and stored in dev
+	filp->private_data = dev; //THe open syscall sets filp to NULL before calling the open method to the driver. By assigning addr, other modules can use it. 
+    /**
+     * TODO: handle open
+     */
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+    /**
+     * TODO: handle release
+     */
+    return 0;
+}
+
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    size_t offset_pos;
+    int kernel_buffer_count = 0;
+    struct aesd_dev *dev = filp->private_data;
+    struct aesd_buffer_entry *kernel_buff;
+
+    PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+
+    mutex_lock(&aesd_device.lock);  //Kernel lock primitive
+
+    kernel_buff = aesd_circular_buffer_find_entry_offset_for_fpos(&dev->aesd_circular_buffer, *f_pos, &offset_pos);
+
+    if (kernel_buff == NULL) //if address was not returned. 
+    {      
+        *f_pos = 0; 
+        goto clean; //goto is not frowned upon in kernel programming, use it to jump to cleanup steps. 
+    }
+    
+    if ((kernel_buff->size - offset_pos) < count) 
+    {
+        *f_pos += (kernel_buff->size - offset_pos);    //Based on the returned offset and size, update f_pos. 
+        kernel_buffer_count = kernel_buff->size - offset_pos;
+    } 
+    
+    else 
+    {
+        *f_pos += count;
+        kernel_buffer_count = count;
+    }
+
+    if (copy_to_user(buf, kernel_buff->buffptr+offset_pos, kernel_buffer_count)) //copying to buf, which is a userspace buffer.
+    {      
+		retval = -EFAULT;
+		goto clean;
+	}
```

Comment:
@suja8742 watch your spacing here and elsewhere

Reply:
Fixed it, thanks Dan.